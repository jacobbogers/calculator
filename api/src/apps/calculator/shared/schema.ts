const divideSchema = {
  type: "object",
  properties: {
    type: {
      type: "string",
      const: "divide",
    },
  },
  required: ["type"],
  additionalProperties: false,
};

const additionSchema = {
  type: "object",
  properties: {
    type: {
      type: "string",
      const: "add",
    },
  },
  required: ["type"],
  additionalProperties: false,
};

const subtractSchema = {
  type: "object",
  properties: {
    type: {
      type: "string",
      const: "subtract",
    },
  },
  required: ["type"],
  additionalProperties: false,
};

const valueSchema = {
  type: "object",
  properties: {
    type: {
      type: "string",
      const: "value",
    },
    payload: {
      type: "number",
    },
  },
  required: ["type", "payload"],
  additionalProperties: false,
};

const finalValueSchema = {
  type: "object",
  properties: {
    type: {
      type: "string",
      const: "value",
    },
    payload: {
      type: "string",
    },
  },
  required: ["type", "payload"],
  additionalProperties: false,
};

const multiplySchema = {
  type: "object",
  properties: {
    type: {
      type: "string",
      const: "multiply",
    },
  },
  required: ["type"],
  additionalProperties: false,
};

export const calculatorRequest = {
  $schema: "http://json-schema.org/draft-07/schema#",
  title: "Calculator Engine Request",
  $id: "https://redspher.com/calculator/request",
  type: "array",
  minItems: 1,
  items: {
    oneOf: [
      multiplySchema,
      divideSchema,
      additionSchema,
      subtractSchema,
      valueSchema,
    ],
  },
};

export const calculatorResponse = {
  $schema: "http://json-schema.org/draft-07/schema#",
  title: "Calculator Engine Response",
  $id: "https://redspher.com/api/calculator/response",
  ...finalValueSchema,
};

export const errorResponse = {
  $schema: "http://json-schema.org/draft-07/schema#",
  title: "Calculator Engine Error Response",
  $id: "https://redspher.com/api/calculator/error",
  type: "object",
  properties: {
    error: {
      type: "object",
      properties: {
        code: { type: "number" },
        message: { type: "string" },
        reqId: { type: "string" },
      },
      required: ["code", "message", "reqId"],
      additionalProperties: false,
    },
    ctx: {
      type: "object",
      additionalProperties: true,
    },
  },
  required: ["error"],
  additionalProperties: false,
};
