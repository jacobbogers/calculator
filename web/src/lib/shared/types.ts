type MultiplyAction = {
    type: 'multiply';
};

type DivideAction = {
    type: 'divide';
};

type AddAction = {
    type: 'add';
};

type SubtractAction = {
    type: 'subtract';
};

type ValueAction = {
    type: 'value';
    payload: number;
};

type FinalvalueAction = {
    type: 'value';
    payload: string;
}

type ServerResponseError = {
    error: {
        code: number;
        message: string;
        reqId: string;
    };
    ctx?: Record<string, unknown>;
};

type MathOperatorActions = MultiplyAction | DivideAction | AddAction | SubtractAction;
type RecordedActionsSansDigit = MathOperatorActions | ValueAction;

export type {
    MultiplyAction,
    DivideAction,
    AddAction,
    SubtractAction,
    ValueAction,
    ServerResponseError,
    MathOperatorActions,
    RecordedActionsSansDigit,
    FinalvalueAction
};
