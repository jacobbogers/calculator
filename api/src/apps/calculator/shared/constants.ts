export const ERR_APP_CALCULATOR_REJECT = 1000;
export const API_REST_PATH_SUFFIX = '/api/v1/calculator/submit';
