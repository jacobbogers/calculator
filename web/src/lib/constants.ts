export const MIME_APPLICATION_JSON = 'application/json';

export const ERR_RESPONSE_ERR_NO_JSON = 9300;
export const ERR_RESPONSE_ERR_NO_VALID_JSON = 9400;

export const ERR_RESPONSE_OK_NO_JSON = 9500;
export const ERR_RESPONSE_OK_NO_VALID_JSON = 9600;
export const ERR_INVALID_REQUEST = 9700;

export const errMsgMap = {
    [ERR_RESPONSE_ERR_NO_JSON]: 'Server responded with an error. The response was not in json format: %text',
    [ERR_RESPONSE_ERR_NO_VALID_JSON]:
        'Server responded with an error. The response was in json, but unknown format: %text',
    [ERR_RESPONSE_OK_NO_JSON]: 'Server responded ok. The response was not in json format: %text',
    [ERR_RESPONSE_OK_NO_VALID_JSON]: 'Server responded ok. THe response was in json, but unknown format: %text',
    [ERR_INVALID_REQUEST]: 'CLIENT ERROR: outgoing request invalid, %text'
};

export function getErr<K extends keyof typeof errMsgMap>(errNo: K, detail: string): string {
    return errMsgMap[errNo].replace('%text', detail);
}
