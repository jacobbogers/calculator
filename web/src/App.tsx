/* vender */
import React, { useCallback, useMemo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import classNames from 'classnames';

/* styles */
import styles from './App.module.scss';

/* components */
import CalculatorBtn from './components/CalculatorBtn';
import KeyBoard from './components/KeyBoard';
import Display from './components/Display';

/* redux */
import type { StoreLayout } from './lib/reducers';

/* misc */
import { createAction } from './lib/actions';

export default function App() {
    const dispatch = useDispatch();
    const isDirty = useSelector<{ tainted: boolean }, boolean>((state) => state.tainted);
    const networkRequestInTransit = useSelector<StoreLayout, boolean>((state) => state.inTransit);

    const safeDispatch = useCallback(
        (...args: Parameters<typeof dispatch>) => {
            if (networkRequestInTransit) {
                return;
            }
            return dispatch(...args);
        },
        [dispatch, networkRequestInTransit]
    );

    const className = useMemo(
        () =>
            classNames({
                [styles.app]: true,
                [styles.wait]: networkRequestInTransit
            }),
        [networkRequestInTransit]
    );

    return (
        <div className={className}>
            <KeyBoard>
                <Display />
                {/* row 1 */}
                <CalculatorBtn disabled darker>
                    Rad
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    Deg
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    x!
                </CalculatorBtn>

                <CalculatorBtn disabled darker>
                    (
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    )
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    %
                </CalculatorBtn>
                {isDirty ? (
                    <CalculatorBtn
                        data-testid="btnClearEntry"
                        darker
                        smaller
                        onClick={() => safeDispatch(createAction('clearEntry'))}
                    >
                        CE
                    </CalculatorBtn>
                ) : (
                    <CalculatorBtn
                        data-testid="btnClearAll"
                        darker
                        smaller
                        onClick={() => safeDispatch(createAction('clearAll'))}
                    >
                        AC
                    </CalculatorBtn>
                )}
                {/* row 2 */}

                <CalculatorBtn disabled darker>
                    Inv
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    sin
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    ln
                </CalculatorBtn>

                <CalculatorBtn data-testid="btn7" onClick={() => safeDispatch(createAction('digit', '7'))}>
                    7
                </CalculatorBtn>
                <CalculatorBtn data-testid="btn8" onClick={() => safeDispatch(createAction('digit', '8'))}>
                    8
                </CalculatorBtn>
                <CalculatorBtn data-testid="btn9" onClick={() => safeDispatch(createAction('digit', '9'))}>
                    9
                </CalculatorBtn>
                <CalculatorBtn
                    data-testid="btnDivide"
                    darker
                    larger
                    onClick={() => safeDispatch(createAction('divide'))}
                >
                    ÷
                </CalculatorBtn>

                {/* row 3 */}

                <CalculatorBtn disabled darker>
                    π
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    cos
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    log
                </CalculatorBtn>

                <CalculatorBtn data-testid="btn4" onClick={() => safeDispatch(createAction('digit', '4'))}>
                    4
                </CalculatorBtn>
                <CalculatorBtn data-testid="btn5" onClick={() => safeDispatch(createAction('digit', '5'))}>
                    5
                </CalculatorBtn>
                <CalculatorBtn data-testid="btn6" onClick={() => safeDispatch(createAction('digit', '6'))}>
                    6
                </CalculatorBtn>
                <CalculatorBtn
                    data-testid="btnMultiply"
                    darker
                    larger
                    onClick={() => safeDispatch(createAction('multiply'))}
                >
                    ×
                </CalculatorBtn>

                {/* row 4 */}

                <CalculatorBtn disabled darker>
                    e
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    tan
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    √
                </CalculatorBtn>

                <CalculatorBtn data-testid="btn1" onClick={() => safeDispatch(createAction('digit', '1'))}>
                    1
                </CalculatorBtn>
                <CalculatorBtn data-testid="btn2" onClick={() => safeDispatch(createAction('digit', '2'))}>
                    2
                </CalculatorBtn>
                <CalculatorBtn data-testid="btn3" onClick={() => safeDispatch(createAction('digit', '3'))}>
                    3
                </CalculatorBtn>
                <CalculatorBtn
                    data-testid="btnSubtract"
                    darker
                    larger
                    onClick={() => safeDispatch(createAction('subtract'))}
                >
                    −
                </CalculatorBtn>

                {/* row 5 */}

                <CalculatorBtn disabled darker>
                    Ans
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    EXP
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    <span>xʸ</span>
                </CalculatorBtn>

                <CalculatorBtn data-testid="btn0" onClick={() => safeDispatch(createAction('digit', '0'))}>
                    0
                </CalculatorBtn>
                <CalculatorBtn data-testid="btnDot" bolder onClick={() => safeDispatch(createAction('digit', '.'))}>
                    .
                </CalculatorBtn>
                <CalculatorBtn
                    data-testid="btnSubmit"
                    bolder
                    larger
                    submit
                    onClick={() => safeDispatch(createAction('result'))}
                >
                    =
                </CalculatorBtn>
                <CalculatorBtn data-testid="btnAdd" darker larger onClick={() => safeDispatch(createAction('add'))}>
                    +
                </CalculatorBtn>
            </KeyBoard>
        </div>
    );
}
