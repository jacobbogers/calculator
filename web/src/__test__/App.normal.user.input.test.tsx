import React from 'react';
import type { Store } from 'redux';
import type { MockResponseInit } from 'jest-fetch-mock';
import { cleanup, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import createStore from '../lib/store';
import App from '../App';

import { API_REST_PATH_SUFFIX } from '../lib/shared/constants';

import { getUIElements, customRender } from './test-helper';
import type { StoreLayout } from '../lib/reducers';

describe('Normal User input', () => {
    let store: Store<StoreLayout>;

    beforeEach(() => {
        store = createStore();
        fetchMock.enableMocks();
    });

    afterEach(() => {
        cleanup();
        fetchMock.resetMocks();
    });
    it('format 1_000_000_000_000 as 1e+12', async () => {
        const { unmount } = customRender(<App />, store);
        const user = userEvent.setup({ delay: 0.4 });
        const { btn0, btn1, commandLine, btnSubmit } = getUIElements(screen);
        await user.click(btn1);
        for (let i = 0; i < 12; i++) {
            await user.click(btn0);
        }
        await user.click(btnSubmit);
        expect(commandLine.textContent).toBe('1e+12');
        unmount();
    });
    it('format 1_943_000_000_000 as 1e+12', async () => {
        const { unmount } = customRender(<App />, store);
        const user = userEvent.setup({ delay: 0.4 });
        const { btn0, btn1, btn9, btn4, btn3, commandLine, btnSubmit } = getUIElements(screen);
        await user.click(btn1);
        await user.click(btn9);
        await user.click(btn4);
        await user.click(btn3);
        for (let i = 0; i < 9; i++) {
            await user.click(btn0);
        }
        await user.click(btnSubmit);
        expect(commandLine.textContent).toBe('1.943e+12');
        unmount();
    });
    it('format 123_456_000_000 is un-adjusted', async () => {
        const { unmount } = customRender(<App />, store);
        const user = userEvent.setup({ delay: 0.4 });
        const { btn0, btn1, btn2, btn3, btn4, btn5, btn6, commandLine, btnSubmit } = getUIElements(screen);
        await user.click(btn1);
        await user.click(btn2);
        await user.click(btn3);
        await user.click(btn4);
        await user.click(btn5);
        await user.click(btn6);
        for (let i = 0; i < 6; i++) {
            await user.click(btn0);
        }
        await user.click(btnSubmit);
        expect(commandLine.textContent).toBe('123456000000');
        unmount();
    });
    it('format "0.123_456_000_9" to "0.123456001"', async () => {
        const { unmount } = customRender(<App />, store);
        const user = userEvent.setup({ delay: 0.4 });
        const { btnDot, btn0, btn1, btn2, btn3, btn4, btn5, btn6, btn9, commandLine, btnSubmit } =
            getUIElements(screen);
        await user.click(btn0);
        await user.click(btnDot);
        await user.click(btn1);
        await user.click(btn2);
        await user.click(btn3);
        await user.click(btn4);
        await user.click(btn5);
        await user.click(btn6);
        for (let i = 0; i < 3; i++) {
            await user.click(btn0);
        }
        await user.click(btn9);
        await user.click(btnSubmit);
        expect(commandLine.textContent).toBe('0.123456001');
        unmount();
    });
    it('format "1_123_456_789.000_9 to become "1_123_456_789"', async () => {
        const { unmount } = customRender(<App />, store);
        const user = userEvent.setup({ delay: 0.4 });
        const { btnDot, btn0, btn1, btn2, btn3, btn4, btn5, btn6, btn7, btn8, btn9, commandLine, btnSubmit } =
            getUIElements(screen);
        await user.click(btn1);
        //
        await user.click(btn1);
        await user.click(btn2);
        await user.click(btn3);
        //
        await user.click(btn4);
        await user.click(btn5);
        await user.click(btn6);
        //
        await user.click(btn7);
        await user.click(btn8);
        await user.click(btn9);
        //
        await user.click(btnDot);
        for (let i = 0; i < 3; i++) {
            await user.click(btn0);
        }
        await user.click(btn9);
        await user.click(btnSubmit);
        expect(commandLine.textContent).toBe('1123456789');
        unmount();
    });
    it('format "0.001/123_456_789_123_456_789" = "8.100000065610001e-22" becomes "8.1000001e-23"', async () => {
        const { unmount } = customRender(<App />, store);
        const user = userEvent.setup({ delay: 0.4 });
        const {
            btnDot,
            btn0,
            btn1,
            btn2,
            btn3,
            btn4,
            btn5,
            btn6,
            btn7,
            btn8,
            btn9,
            btnDivide,
            commandLine,
            btnSubmit
        } = getUIElements(screen);

        await fetchMock.mockResponse((request: Request) => {
            return Promise.resolve<MockResponseInit>({
                init: {
                    headers: {
                        'content-type': 'application/json'
                    },
                    status: 200,
                    statusText: 'ok'
                },
                body: '{ "type": "value", "payload": 8.100000065610001e-23 }'
            });
        });

        await user.click(btnDot);
        for (let i = 0; i < 4; i++) {
            await user.click(btn0);
        }
        await user.click(btn1);
        // entered .0001
        await user.click(btnDivide);
        //
        for (let i = 0; i < 2; i++) {
            await user.click(btn1);
            await user.click(btn2);
            await user.click(btn3);
            //
            await user.click(btn4);
            await user.click(btn5);
            await user.click(btn6);
            //
            await user.click(btn7);
            await user.click(btn8);
            await user.click(btn9);
        }
        await user.click(btnSubmit);
        expect(commandLine.textContent).toBe('8.1000001e-23');
        const { commandLine: cmdLineStore } = store.getState();
        expect(cmdLineStore).toEqual([{ type: 'value', payload: 8.100000065610001e-23 }]);
        unmount();
    });
    it('multiply 2 negative numbers', async () => {
        const { unmount } = customRender(<App />, store);
        const user = userEvent.setup({ delay: 0.4 });

        const { btnSubtract, btn9, btn5, btnDot, btn2, btnMultiply, btn6, btnSubmit, commandLine } =
            getUIElements(screen);

        // -92.5x-.6 =
        await user.click(btnSubtract);
        await user.click(btn9);
        await user.click(btn2);
        await user.click(btnDot);
        await user.click(btn5);
        await user.click(btnMultiply);
        await user.click(btnSubtract);
        await user.click(btnDot);
        await user.click(btn6);

        expect(commandLine.textContent).toEqual('-92.5 × -.6');

        await fetchMock.mockResponse((request: Request) => {
            return Promise.resolve<MockResponseInit>({
                init: {
                    headers: {
                        'content-type': 'application/json'
                    },
                    status: 200,
                    statusText: 'ok'
                },
                body: '{ "type": "value", "payload": 55.5 }'
            });
        });

        await user.click(btnSubmit);
        expect(commandLine.textContent).toEqual('55.5');
        unmount();
    });
    it('user journey: Entering "9 x 5.2", 2x click "=" button, x 34 =', async () => {
        const { unmount } = customRender(<App />, store);
        const user = userEvent.setup({ delay: 0.4 });
        // get buttons to emulate keystrokes
        const { btn9, btn5, btnDot, btn2, btn6, btnSubmit, btnDivide, btnMultiply, commandLine, lastHistoryLine } =
            getUIElements(screen);

        expect(commandLine.textContent).toBe('0');
        expect(lastHistoryLine.textContent).toBe('');

        // first key press
        await user.click(btn9);

        // after the first keypress the historyLine is now "Ans =  0"
        expect(lastHistoryLine.textContent).toBe('Ans = 0');

        await user.click(btnMultiply);
        await user.click(btn5);
        await user.click(btnDot);
        await user.click(btn2);

        expect(store.getState()).toMatchSnapshot();
        expect(commandLine.textContent).toBe('9 × 5.2');

        await fetchMock.mockResponse((request: Request) => {
            return Promise.resolve<MockResponseInit>({
                init: {
                    headers: {
                        'content-type': 'application/json'
                    }
                },
                body: JSON.stringify({
                    type: 'value',
                    payload: 46.800000000000004
                })
            });
        });

        await user.click(btnSubmit);
        expect(fetchMock.mock.lastCall).toEqual([
            API_REST_PATH_SUFFIX,
            {
                method: 'POST',
                mode: 'same-origin',
                credentials: 'same-origin',
                redirect: 'follow',
                keepalive: true,
                body: JSON.stringify([
                    { type: 'value', payload: 9 },
                    { type: 'multiply' },
                    { type: 'value', payload: 5.2 }
                ]),
                headers: {
                    accept: 'application/json',
                    'content-type': 'application/json'
                }
            }
        ]);
        expect(store.getState()).toMatchSnapshot();
        expect(commandLine.textContent).toBe('46.8');
        expect(lastHistoryLine.textContent).toBe('9 × 5.2 =');

        // CE buttons should be displaced, AC button should not be rendered
        expect(screen.queryByTestId('btnclearEntry')).toBeNull();
        // throws if it does not exist, this is an assert by itself
        screen.getByTestId('btnClearAll');

        // user clicks = for second time
        await user.click(btnSubmit);
        expect(commandLine.textContent).toBe('46.8');
        expect(lastHistoryLine.textContent).toBe('9 × 5.2 =');

        // continue with calculation using returned result
        await user.click(btnDivide);
        await user.click(btnDot);
        await user.click(btn6);
        expect(store.getState()).toMatchSnapshot();
        expect(commandLine.textContent).toBe('46.8 ÷ .6');

        await fetchMock.mockResponse((request: Request) => {
            return Promise.resolve<MockResponseInit>({
                init: {
                    headers: {
                        'content-type': 'application/json'
                    }
                },
                body: JSON.stringify({
                    type: 'value',
                    payload: 78.00000000000001
                })
            });
        });

        await user.click(btnSubmit);

        expect(commandLine.textContent).toBe('78');
        expect(lastHistoryLine.textContent).toBe('46.8 ÷ .6 =');
        unmount();
    });
});
