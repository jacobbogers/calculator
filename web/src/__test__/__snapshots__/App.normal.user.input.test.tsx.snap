// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Normal User input > user journey: Entering "9 x 5.2", 2x click "=" button, x 34 = 1`] = `
{
  "changeOnReadyOrError": 0,
  "commandLine": [
    {
      "payload": "9",
      "type": "digit",
    },
    {
      "type": "multiply",
    },
    {
      "payload": "5",
      "type": "digit",
    },
    {
      "payload": ".",
      "type": "digit",
    },
    {
      "payload": "2",
      "type": "digit",
    },
  ],
  "history": [],
  "inTransit": false,
  "serverError": {
    "error": undefined,
  },
  "tainted": true,
}
`;

exports[`Normal User input > user journey: Entering "9 x 5.2", 2x click "=" button, x 34 = 2`] = `
{
  "changeOnReadyOrError": 1,
  "commandLine": [
    {
      "payload": 46.800000000000004,
      "type": "value",
    },
  ],
  "history": [
    {
      "keyedInExpression": [
        {
          "payload": "9",
          "type": "digit",
        },
        {
          "type": "multiply",
        },
        {
          "payload": "5",
          "type": "digit",
        },
        {
          "payload": ".",
          "type": "digit",
        },
        {
          "payload": "2",
          "type": "digit",
        },
      ],
      "result": {
        "payload": 46.800000000000004,
        "type": "value",
      },
    },
  ],
  "inTransit": false,
  "serverError": {},
  "tainted": false,
}
`;

exports[`Normal User input > user journey: Entering "9 x 5.2", 2x click "=" button, x 34 = 3`] = `
{
  "changeOnReadyOrError": 2,
  "commandLine": [
    {
      "payload": 46.800000000000004,
      "type": "value",
    },
    {
      "type": "divide",
    },
    {
      "payload": ".",
      "type": "digit",
    },
    {
      "payload": "6",
      "type": "digit",
    },
  ],
  "history": [
    {
      "keyedInExpression": [
        {
          "payload": "9",
          "type": "digit",
        },
        {
          "type": "multiply",
        },
        {
          "payload": "5",
          "type": "digit",
        },
        {
          "payload": ".",
          "type": "digit",
        },
        {
          "payload": "2",
          "type": "digit",
        },
      ],
      "result": {
        "payload": 46.800000000000004,
        "type": "value",
      },
    },
  ],
  "inTransit": false,
  "serverError": {},
  "tainted": true,
}
`;
